cnd/src/network/peer_tracker.rs

```diff
@@ -9,24 +9,63 @@ use libp2p::{
     Multiaddr, PeerId,
 };
 use std::{
-    collections::{hash_map::Entry, HashMap},
+    collections::{hash_map::Entry, HashMap, VecDeque},
     task::Poll,
 };
 
 /// A NetworkBehaviour that tracks connections to other peers.
 #[derive(Default, Debug)]
 pub struct PeerTracker {
     connected_peers: HashMap<PeerId, Vec<Multiaddr>>,
-    address_hints: HashMap<PeerId, Multiaddr>,
+    address_hints: HashMap<PeerId, VecDeque<Multiaddr>>,
 }
 
 impl PeerTracker {
     pub fn connected_peers(&self) -> impl Iterator<Item = (PeerId, Vec<Multiaddr>)> {
         self.connected_peers.clone().into_iter()
     }
 
-    pub fn add_address_hint(&mut self, id: PeerId, addr: Multiaddr) -> Option<Multiaddr> {
-        self.address_hints.insert(id, addr)
+    /// Adds an address hint for the given peer id. The added address is
+    /// considered most recent and hence is added at the start of the list.
+    pub fn add_address_hint(&mut self, id: PeerId, addr: Multiaddr) {
+        let old_addresses = self.address_hints.get_mut(&id);
+
+        match old_addresses {
+            None => {
+                let mut hints = VecDeque::new();
+                hints.push_back(addr);
+                self.address_hints.insert(id, hints);
+            }
+            Some(hints) => {
+                // Put the new address at the start of the list.
+                hints.push_front(addr);
+            }
+        }
+    }
+
+    /// Adds address hints for the given peer id. The added addresses are
+    /// considered most recent and hence are added at the start of the list.
+    pub fn add_address_hints(&mut self, id: PeerId, mut addresses: Vec<Multiaddr>) {
+        let old_addresses = self.address_hints.get_mut(&id);
+
+        match old_addresses {
+            None => {
+                let mut hints = VecDeque::new();
+                for _ in 0..addresses.len() {
+                    // Need to keep same order than the vec
+                    let addr = addresses.pop().expect("We loop on the length");
+                    hints.push_front(addr);
+                }
```

Comment:
Does this do the trick as well?

```suggestion
                hints.extend(addresses);
```